proxy_request.c

```diff
@@ -753,7 +767,19 @@ int mcplib_request_flag_token(lua_State *L) {
                 if (vlen > 1) {
                     // strip the flag off the token and return.
                     lua_pushlstring(L, s+1, vlen-1);
-                    return 2;
+                    ret = 2;
+                }
+
+                // Have something to replace the flag/token with.
+                if (replace) {
```

Comment:
Thanks for the quick drop @dormando !

One of the things I noticed in both token() and flag_token() is that after changing the value of either of them, we set `r->was_modified = true`

However, we don't change the actual metadata of the original request itself.

For example, if I do the following:
```
      print(r:flag_token("C"))
      r:flag_token("C","")   -- blank out the CAS token
      print(r:flag_token("C"))
```

I get the following output:
```
true	6
true	6
```

For our purposes, we don't need to modify the original request more than once (yet).

But from an API guarantees POV, what do you think would be the right approach here?

Reply:
> Debating if I should move more of the behavior of r:token() over where it can return updated tokens, or if I document it as a "fast" function that only handles the original request string.

Is this what you were referring to here? ^

Reply:
Yes, that's what I'm referring to. The token accessors are optimized so you can make a bunch of modifications without having to re-write the request string over and over. `was_modified` is a hint for the request attach function to flatten the new request query.

I'm not really sure what the right answer is; I'm tempted to leave it like this for now and change it soonish. I had some ideas for how to speed up request rewriting such that it might not matter anymore. ie; now that the endcap works you can fast-copy before/after a token, so it ends up being extra memcpy's instead of all the lua table fiddling. That'll either come out break even or slightly ahead due to fewer allocations.

but I'd like to avoid doing that this week.

Reply:
So; yeah the difference here is `r:token()` is always checking the lua table to see if a token's been overwritten. I can add that but we lose some "fast-pathy-ness".

Probably _should_ do it for expectations but not super sure it's necessary.

`local (exists, old) = r:flag_token("C", "Cnew")`

in this case old is not "new", but is whatever the previous token was. If a use case needs the previous token you have it there, and if you need the new one a second time you already have it as you just set it.

If you issue a request, IIRC it flattens everything. So if you modify a request with `r:flag_token`, issue it to a backend, then modify it again, you'll get the new data.

From there it gets kinda complex unfortunately.

Reply:
Ack, that's fair. For our purposes, this works for now.

The only thing I'd add here is to make sure that the APIs are documented with their appropriate design choices/limitations, so that it's clear to other folks adopting it.

Reply:
I'll update the docs on release and make a note about a potential behavior change, thanks!